### PR TITLE
add onnx support for deberta and debertav2

### PR DESCRIPTION
**Pull request by sam-h-bean**
# What does this PR do?

Details: Add ONNX Support for DeBERTa and DeBERTaV2.

Issue: https://github.com/huggingface/optimum/issues/207

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?
@JingyaHuang @ChainYo

### CODE REVIEW

**Comment by sam-h-bean** [comment]
@JingyaHuang I do not have the ability to rerun the failed CI/CD checks. The failures seem unrelated to my changes so unsure what changes I need to be making

File: src/transformers/onnx/features.py
Comment (ChainYo) [inline src/transformers/onnx/features.py:208]: I'm not sure if there is `causal-lm` as a supported feature for Debertav2.

**Comment by ChainYo** [comment]
Hello @sam-h-bean, thanks for the PR; it looks nice!

File: src/transformers/onnx/features.py
Comment (JingyaHuang) [reply src/transformers/onnx/features.py:208]: To confirm `DebertaV2ForCausalLM` is not implemented, shall remove `causal-lm`.

**Review by JingyaHuang — COMMENTED** [review]
Hi @sam-h-bean, thanks for contributing! The `OnnxConfig` of DeBERTa-V2 looks good to me. Can you also add it to the test [tests/onnx/test_onnx_v2.py](https://github.com/huggingface/transformers/blob/main/tests/onnx/test_onnx_v2.py#L176)  and run
```
RUN_SLOW=1 pytest tests/onnx/test_onnx_v2.py -s -k "deberta_v2"
```
to confirm that the export goes well?

File: src/transformers/onnx/features.py
Comment (sam-h-bean) [reply src/transformers/onnx/features.py:208]: done

**Comment by sam-h-bean** [comment]
I am seeing that [deberta v2 is leveraging some symbolic library using onnx already](https://github.com/huggingface/transformers/blob/main/src/transformers/models/deberta_v2/modeling_deberta_v2.py#L124-L137). Line 135 is where the crash is happening now

**Comment by JingyaHuang** [comment]
Hi @sam-h-bean, the symbolic function was added intentionally(check #14013) to support ONNX export as XSoftmax is not a natively supported onnx op. Can you try to export Deberta-V2 with an upper opset, eg.  with `opset=15(onnx>=1.10.0)`
```
onnx_inputs, onnx_outputs = export(preprocessor, model, onnx_config, 15, Path(output.name), device=device)
```
And maybe find the minimal opset if possible(currently we set [default it to 11](https://github.com/huggingface/transformers/blob/main/src/transformers/onnx/config.py#L40)).

Keep me posted if it works, then we can make the necessary change on the export.

File: src/transformers/models/deberta_v2/configuration_deberta_v2.py
Comment (sam-h-bean) [reply src/transformers/models/deberta_v2/configuration_deberta_v2.py:160]: @JingyaHuang All opsets 12-15 fail with

```
FAILED tests/onnx/test_onnx_v2.py::OnnxExportTestCaseV2::test_pytorch_export_043_deberta_v2_default - AssertionError: deberta-v2, default -> ONNX export failed: Couldn't export Python operator XSoftmax
```

I tried replacing the actual XSoftmax symbolic method with upper opsets as well and none of them have support for both masked_fill and softmax functions. I am wondering if this is the issue. Maybe this function needs to be rewritten without using opset 9?

**Comment by JingyaHuang** [comment]
@sam-h-bean To confirm, can you try this snippet:
```python
from collections import OrderedDict
from typing import Mapping
from pathlib import Path
from transformers.onnx import export
from transformers.onnx import OnnxConfig
from transformers import AutoTokenizer, AutoModel, AutoConfig

class DebertaV2OnnxConfig(OnnxConfig):
    @property
    def inputs(self) -> Mapping[str, Mapping[int, str]]:
        if self.task == "multiple-choice":
            dynamic_axis = {0: "batch", 1: "choice", 2: "sequence"}
        else:
            dynamic_axis = {0: "batch", 1: "sequence"}
        return OrderedDict(
            [
                ("input_ids", dynamic_axis),
                ("attention_mask", dynamic_axis),
            ]
        )
    @property
    def default_onnx_opset(self) -> int:
        return 15

config = AutoConfig.from_pretrained("microsoft/deberta-v3-large")
base_model = AutoModel.from_pretrained("microsoft/deberta-v3-large")
tokenizer = AutoTokenizer.from_pretrained("microsoft/deberta-v3-large")
onnx_config = DebertaV2OnnxConfig(config)
onnx_path = Path("deberta.onnx")
onnx_inputs, onnx_outputs = export(tokenizer, base_model, onnx_config, onnx_config.default_onnx_opset, onnx_path)
print(onnx_config.default_onnx_opset)
```

File: src/transformers/models/deberta_v2/configuration_deberta_v2.py
Comment (JingyaHuang) [reply src/transformers/models/deberta_v2/configuration_deberta_v2.py:160]: There shall not be anything to change in the modeling of deberta-v2. Just tried exporting with your customized OnnxConfig, it should work by correctly setting the default opset.

**Comment by sam-h-bean** [comment]
@JingyaHuang I tried the export in a try/except as follows

```python
try:
     onnx_inputs, onnx_outputs = export(tokenizer, base_model, onnx_config, onnx_config.default_onnx_opset, onnx_path)
 except (RuntimeError, ValueError) as e:
     print(onnx_config.default_onnx_opset)
```

And it prints 15

**Comment by michaelbenayoun** [comment]
Hi @sam-h-bean, 

I merged a PR a few days ago changing deberta symbolic function #17539 (to follow what is planned to be done [here](https://github.com/huggingface/transformers/pull/17306#issuecomment-1144846653)), that seems to break the ONNX export.

I think that the culprit here is the  `torch.finfo(some_tensor.dtype)`  part. Let me try to find a workaround.

Pinging @ydshieh to make him aware of the issue.

**Comment by sam-h-bean** [comment]
@michaelbenayoun @ydshieh this PR now contains a fix for the symbolic function

File: tests/onnx/test_onnx_v2.py
Comment (sam-h-bean) [inline tests/onnx/test_onnx_v2.py:268]: @JingyaHuang A weird thing about the DeBERTa model is that it is the only model that by default takes token_type_ids in its forward method (the tokenizer returns this in the dictionary by default) which the model config by default sets the vocab size to 0 (this basically ignores that value in the forward method).

This has a few downstream side-effects, the ONNX graph perceives this as meaning the model by default only takes 2 inputs which causes the test to fail because we attempt to pass all 3 items in the ONNX input. What I have done is in the ONNX config I override the value to be 2 so that the graph is constructed taking 3 inputs (the default for the tokenizer but not the model) and had to hack in this statement to initialize the model config to expect token_type_ids as well.

This feels pretty bad. I think that the tokenizer should by default only return 2 items unless the model config has type_vocab_size > 0. However, this would be a larger change to the DeBERTa classes and I was unsure if that was desired.

File: src/transformers/models/deberta_v2/modeling_deberta_v2.py
Comment (sam-h-bean) [inline src/transformers/models/deberta_v2/modeling_deberta_v2.py:135]: This is a bug that was introduced recently. self has no dtype

**Comment by ydshieh** [comment]
I have very little knowledge on this onnx thing. But what I can add here is: we need to understand what `self` actually is in
```def symbolic(g, self, mask, dim):
```
My quick search (in a previous PR) shows it should be some input to the model. And if it is a `torch.Tensor`, it should have `dtype`. 

- Maybe it is not input to model?
- Maybe it is, but for some reason, it is not `torch.Tensor` (for example some symbolic tensor etc - I have no idea)

I suggest to set some break points, launch a code that will call this method, and investigate what `self` is.
Please don't merge with `float32.dtype` without any further investigate, thank you 🙏

**Comment by sam-h-bean** [comment]
> I have very little knowledge on this onnx thing. But what I can add here is: we need to understand what `self` actually is in
> 
> ```python
> def symbolic(g, self, mask, dim):
> ```
> 
> My quick search (in a previous PR) shows it should be some input to the model. And if it is a `torch.Tensor`, it should have `dtype`.
> 
>     * Maybe it is not input to model?
> 
>     * Maybe it is, but for some reason, it is not `torch.Tensor` (for example some symbolic tensor etc - I have no idea)
> 
> 
> I suggest to set some break points, launch a code that will call this method, and investigate what `self` is. Please don't merge this PR with `float32.dtype` without any further investigate first, thank you 🙏

This code previously was 

```python
output = masked_fill(g, self, r_mask, g.op("Constant", value_t=torch.tensor(float("-inf"))))
```

@ydshieh  Then it was changed to this self reference which broke the code. This is now using the built-in torch -inf so the type will be the same. I can also just change it back to `torch.tensor(float("-inf"))` if you would prefer but I imagine this won't make a great deal of difference in practice

**Comment by ydshieh** [comment]
We would like to have the same logic applies to the usual model as well as in ONNX here. For the usual PyTorch models, we use the `dtype` of the inputs (which are torch tensors) to determine the large negative values.

I understand we also need to make ONNX work. Let's dive deeper next week.

In the meantime, could you provide the code snippet you used to investigate these values in your above comment?
(I didn't check all your previous comments , so probably it is there already)

If you have some more time for now, would you mind to check what type `self` is, say `type(self)`. You can also check `dir(self)` to see what kind of attributes it has.

It looks like some kind of symbolic tensor (in some sense).

**Comment by sam-h-bean** [comment]
Screen Shot 2022-06-10 at 6 08 57 PM

It seems like some compiled C tensor. 

You can see it has a type but no dtype below
Screen Shot 2022-06-10 at 6 09 47 PM

**Comment by sam-h-bean** [comment]
@ydshieh Sorry I did not see your request to get the debugging steps. I hacked the onnx test [here](https://github.com/huggingface/transformers/blob/main/tests/onnx/test_onnx_v2.py#L304) to look like the following
```python
    # @parameterized.expand(_get_models_to_test(PYTORCH_EXPORT_MODELS))
    # @slow
    @require_torch
    @require_vision
    @require_rjieba
    def test_pytorch_export(self):
        self._onnx_export("test_name", "name", "microsoft/deberta-v2-xlarge", "default", DebertaV2OnnxConfig)
```

Then I imported the requisite `DebertaV2OnnxConfig` and set a breakpoint in the  PyCharm debugger [here](https://github.com/huggingface/transformers/blob/main/src/transformers/models/deberta_v2/modeling_deberta_v2.py#L135). Then I inspect in the debugger.

**Comment by ydshieh** [comment]
@sam-h-bean 

Thank you for the code snippet. After some inspection, could you try
```python
self.type().dtype()
```
Screenshot 2022-06-11 120746

**Comment by sam-h-bean** [comment]
> @sam-h-bean
> 
> Thank you for the code snippet. After some inspection, could you try
> 
> ```python
> self.type().dtype()
> ```
> 
> Screenshot 2022-06-11 120746

That works! Thank you for the help, I agree deriving the type directly from the output is much more clean and robust

File: tests/onnx/test_onnx_v2.py
Comment (JingyaHuang) [reply tests/onnx/test_onnx_v2.py:267]: Hi @sam-h-bean whether the exported ONNX model needs 2 or 3 inputs depends on whether you've included `token_type_ids` in the custom OnnxConfig, as the dummy inputs used for tracing will [look into the custom OnnxConfig](https://github.com/huggingface/transformers/blob/v4.19.4/src/transformers/onnx/convert.py#L139) to decide what generated inputs will be considered for export. 

As the tokenizer of DeBERTa will return `token_type_ids` by default, we suggest including `token_type_ids` into its OnnxConfig even if it is optional(it will be [ignored](https://github.com/huggingface/transformers/blob/main/src/transformers/models/deberta_v2/modeling_deberta_v2.py#L823) if `type_vocab_size` is set to 0)

**Review by JingyaHuang — COMMENTED** [review]
@sam-h-bean Thanks for fixing the DeBERTa models, sorry for my misleading comments last week as I tested your custom OnnxConfig based on a released version of transformers. Can you run the slow test once again to confirm that nothing is broken?

File: src/transformers/models/deberta_v2/configuration_deberta_v2.py
Comment (ydshieh) [reply src/transformers/models/deberta_v2/configuration_deberta_v2.py:165]: @JingyaHuang I am mostly here for the part regarding `torch.tensor(torch.finfo(self.dtype).min))`, and haven't follow the discussion about `token_type_ids`. Do you intend to tag me here?

File: tests/onnx/test_onnx_v2.py
Comment (sam-h-bean) [reply tests/onnx/test_onnx_v2.py:267]: Yeah I can do that but here we still need to set the `AutoConfig` value some way because the token type embedding is initialized on creation here. So if I just have it overwritten in the onnx config it fails for there being no token embedding layer since by default there is none. So how should I change the config here dynamically?

File: src/transformers/models/deberta_v2/configuration_deberta_v2.py
Comment (sam-h-bean) [reply src/transformers/models/deberta_v2/configuration_deberta_v2.py:165]: OK you were right that this was not needed here. I have removed.

File: tests/onnx/test_onnx_v2.py
Comment (sam-h-bean) [inline tests/onnx/test_onnx_v2.py:267]:
```suggestion
        if model_name == "microsoft/deberta-v2-xlarge":
            config = AutoConfig.from_pretrained(model_name, type_vocab_size=2)
        else:
            config = AutoConfig.from_pretrained(model_name)
```

Comment (sam-h-bean) [reply tests/onnx/test_onnx_v2.py:267]: @JingyaHuang Is this how you would prefer I have type vocab size on for the test?

Comment (JingyaHuang) [reply tests/onnx/test_onnx_v2.py:267]: @sam-h-bean You can do it like that for your own test, to confirm that the export works and in the exported IR there will be the embedding of token type. However, in the transformers test it looks a little bit hacky to set a conditional statement for deberta models only, as deberta is not the only one can have `token_type_ids`. I would prefer that you remove these line from the test in transformers.

Comment (JingyaHuang) [inline tests/onnx/test_onnx_v2.py:267]:
```suggestion
```

Comment (sam-h-bean) [reply tests/onnx/test_onnx_v2.py]: Removed!

**Review by JingyaHuang — COMMENTED** [review]
@sam-h-bean, thanks for the modifications! Can you do the last check by running the slow tests?

**Comment by JingyaHuang** [comment]
Pinging @michaelbenayoun for a review, especially for the modifications on the symbolic function.

**Comment by sam-h-bean** [comment]
> @sam-h-bean, thanks for the modifications! Can you do the last check by running the slow tests?

@JingyaHuang The test will not pass as-is. We either need to modify the tokenizer to return only `input_ids` and `attention_mask` when `type_vocab_size` = 0, or we need to run the test to modify the config we initialize the ONNX config to have a non-zero `type_vocab_size`. This is what I was trying to get across in my comment. By default the ONNX graph will be made expecting 2 inputs since the behavior is to ignore `type_token_ids` in DeBERTa. This is contrary to how Electra works which by default has `type_token_ids`

**Comment by JingyaHuang** [comment]
> RUN_SLOW=1 pytest tests/onnx/test_onnx_v2.py -s -k "deberta_v2"

Hi @sam-h-bean, my mistake, for sure it won't pass. To follow the default setting of the tokenizer, and to have the test passed without adding lines specific to DeBERTa, I would prefer to tailor the inputs according to the `type_vocab_size` to satisfy both scenarios.

File: src/transformers/models/deberta_v2/configuration_deberta_v2.py
Comment (JingyaHuang) [inline src/transformers/models/deberta_v2/configuration_deberta_v2.py:156]:
```suggestion
        if self._config.type_vocab_size > 0:
            return OrderedDict(
                [
                    ("input_ids", dynamic_axis),
                    ("attention_mask", dynamic_axis),
                    ("token_type_ids", dynamic_axis)
                ]
            )
        else:
            return OrderedDict(
                [
                    ("input_ids", dynamic_axis),
                    ("attention_mask", dynamic_axis),
                ]
            )
```
WDYT

Comment (sam-h-bean) [reply src/transformers/models/deberta_v2/configuration_deberta_v2.py]: I'm trying it out now!

Comment (sam-h-bean) [reply src/transformers/models/deberta_v2/configuration_deberta_v2.py]: If I add a value override like

```python
@property
  def values_override(self) -> Optional[Mapping[str, Any]]:
    if hasattr(self._config, "type_vocab_size"):
       return {"type_vocab_size": 2}
```

Then the actual forward pass fails because the underlying model is still initialized without the token type embeddings like

```bash
FAILED tests/onnx/test_onnx_v2.py::OnnxExportTestCaseV2::test_pytorch_export_043_deberta_v2_default - AttributeError: 'DebertaV2Embeddings' object has no attribute 'token_type_embeddings'
```

Comment (sam-h-bean) [reply src/transformers/models/deberta_v2/configuration_deberta_v2.py]: @JingyaHuang The tests pass again with an overridden `generate_dummy_inputs` where if the type vocab size is 0 we delete `type_token_ids` from the dummy inputs. 

**BE WARNED**:
This means that when someone (like me) is actually using this ONNX graph in production they will need to also cleanse the `type_token_ids` from the output of their tokenizer before passing it to the model. There might also be downstream consequences when I work on the quantization pieces for the optimum library.

Comment (JingyaHuang) [reply src/transformers/models/deberta_v2/configuration_deberta_v2.py]: @sam-h-bean 
For the first concern on deleting the `token_type_ids`, as `onnxruntime.InferenceSession` only accepts valid inputs, if the exported ONNX doesn't take `token_type_ids` as input(the case `type_vocab_size=0`), you can only delete the invalid inputs before sending it to the inference session. 

I find it might be interesting to include a filter in [ORTModel](https://github.com/huggingface/optimum/blob/main/optimum/onnxruntime/modeling_ort.py#L317), gently pining @philschmid. Does it make sense to consider this kind of edge case and add a filter into the forward pass to filter out invalid inputs?

Comment (sam-h-bean) [reply src/transformers/models/deberta_v2/configuration_deberta_v2.py]: @JingyaHuang So what does that change look like in practice?

Comment (JingyaHuang) [reply src/transformers/models/deberta_v2/configuration_deberta_v2.py]: For the second concern on the quantization, there is __no impact__ if you are doing __dynamic quantization__. 

If you want to do static quantization and have some worries about the computed range of the calibration dataset. There is __no impact on the performance__, but if the ONNX IR doesn't have `token_type_ids` as input, and there are `token_type_ids` in the calibration dataset it will throw out an error. To avoid this, please use [`get_calibration_dataset`](https://github.com/huggingface/optimum/blob/main/optimum/onnxruntime/quantization.py#L375) which clean the calibration dataset by applying [`clean_calibration_dataset`](https://github.com/huggingface/optimum/blob/main/optimum/onnxruntime/quantization.py#L375).

P.S. I am going to try a nit on tracing the ONNX graph to see if we can avoid so many changes.

Comment (sam-h-bean) [reply src/transformers/models/deberta_v2/configuration_deberta_v2.py]: OK sounds good! So what else needs to happen on this PR to see an approval?

**Comment by sam-h-bean** [comment]
Here is the full test suite passing

```bash
10571 ± RUN_SLOW=1 pytest tests/onnx/test_onnx_v2.py -k "deberta" -v                                                                                                                          ⏎
===================================================================================== test session starts ======================================================================================
platform darwin -- Python 3.9.10, pytest-7.1.2, pluggy-1.0.0 -- /Users/marklar/workspace/transformers/venv/bin/python3
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/Users/marklar/workspace/transformers/.hypothesis/examples')
rootdir: /Users/marklar/workspace/transformers, configfile: setup.cfg
plugins: xdist-2.5.0, forked-1.4.0, timeout-2.1.0, hypothesis-6.47.0, dash-2.5.0
collected 367 items / 345 deselected / 22 selected                                                                                                                                             

tests/onnx/test_onnx_v2.py::OnnxExportTestCaseV2::test_pytorch_export_043_deberta_v2_default PASSED                                                                                      [  4%]
tests/onnx/test_onnx_v2.py::OnnxExportTestCaseV2::test_pytorch_export_044_deberta_v2_masked_lm PASSED                                                                                    [  9%]
tests/onnx/test_onnx_v2.py::OnnxExportTestCaseV2::test_pytorch_export_045_deberta_v2_multiple_choice PASSED                                                                              [ 13%]
tests/onnx/test_onnx_v2.py::OnnxExportTestCaseV2::test_pytorch_export_046_deberta_v2_question_answering PASSED                                                                           [ 18%]
tests/onnx/test_onnx_v2.py::OnnxExportTestCaseV2::test_pytorch_export_047_deberta_v2_sequence_classification PASSED                                                                      [ 22%]
tests/onnx/test_onnx_v2.py::OnnxExportTestCaseV2::test_pytorch_export_048_deberta_v2_token_classification PASSED                                                                         [ 27%]
tests/onnx/test_onnx_v2.py::OnnxExportTestCaseV2::test_pytorch_export_049_deberta_default PASSED                                                                                         [ 31%]
tests/onnx/test_onnx_v2.py::OnnxExportTestCaseV2::test_pytorch_export_050_deberta_masked_lm PASSED                                                                                       [ 36%]
tests/onnx/test_onnx_v2.py::OnnxExportTestCaseV2::test_pytorch_export_051_deberta_question_answering PASSED                                                                              [ 40%]
tests/onnx/test_onnx_v2.py::OnnxExportTestCaseV2::test_pytorch_export_052_deberta_sequence_classification PASSED                                                                         [ 45%]
tests/onnx/test_onnx_v2.py::OnnxExportTestCaseV2::test_pytorch_export_053_deberta_token_classification PASSED                                                                            [ 50%]
tests/onnx/test_onnx_v2.py::OnnxExportTestCaseV2::test_pytorch_export_on_cuda_043_deberta_v2_default PASSED                                                                              [ 54%]
tests/onnx/test_onnx_v2.py::OnnxExportTestCaseV2::test_pytorch_export_on_cuda_044_deberta_v2_masked_lm PASSED                                                                            [ 59%]
tests/onnx/test_onnx_v2.py::OnnxExportTestCaseV2::test_pytorch_export_on_cuda_045_deberta_v2_multiple_choice PASSED                                                                      [ 63%]
tests/onnx/test_onnx_v2.py::OnnxExportTestCaseV2::test_pytorch_export_on_cuda_046_deberta_v2_question_answering PASSED                                                                   [ 68%]
tests/onnx/test_onnx_v2.py::OnnxExportTestCaseV2::test_pytorch_export_on_cuda_047_deberta_v2_sequence_classification PASSED                                                              [ 72%]
tests/onnx/test_onnx_v2.py::OnnxExportTestCaseV2::test_pytorch_export_on_cuda_048_deberta_v2_token_classification PASSED                                                                 [ 77%]
tests/onnx/test_onnx_v2.py::OnnxExportTestCaseV2::test_pytorch_export_on_cuda_049_deberta_default PASSED                                                                                 [ 81%]
tests/onnx/test_onnx_v2.py::OnnxExportTestCaseV2::test_pytorch_export_on_cuda_050_deberta_masked_lm PASSED                                                                               [ 86%]
tests/onnx/test_onnx_v2.py::OnnxExportTestCaseV2::test_pytorch_export_on_cuda_051_deberta_question_answering PASSED                                                                      [ 90%]
tests/onnx/test_onnx_v2.py::OnnxExportTestCaseV2::test_pytorch_export_on_cuda_052_deberta_sequence_classification PASSED                                                                 [ 95%]
tests/onnx/test_onnx_v2.py::OnnxExportTestCaseV2::test_pytorch_export_on_cuda_053_deberta_token_classification PASSED                                                                    [100%]
```

**Review by michaelbenayoun — APPROVED** [review]
Nice work, thanks a lot!!

File: src/transformers/models/deberta/configuration_deberta.py
Comment (JingyaHuang) [inline src/transformers/models/deberta/configuration_deberta.py:186]:
```suggestion
        dummy_inputs = super().generate_dummy_inputs(preprocessor=preprocessor, framework=framework)
        if self._config.type_vocab_size == 0 and "token_type_ids" in dummy_inputs:
            del dummy_inputs["token_type_ids"]
        return dummy_inputs
```

File: src/transformers/models/deberta_v2/configuration_deberta_v2.py
Comment (JingyaHuang) [inline src/transformers/models/deberta_v2/configuration_deberta_v2.py:187]:
```suggestion
        dummy_inputs = super().generate_dummy_inputs(preprocessor=preprocessor, framework=framework)
        if self._config.type_vocab_size == 0 and "token_type_ids" in dummy_inputs:
            del dummy_inputs["token_type_ids"]
        return dummy_inputs
```
Just to keep the consistency of variable name.

**Review by JingyaHuang — APPROVED** [review]
Looks great, thanks for iterating it! 

It probably makes sense to filter out the invalid inputs and I opened a PR on Optimum. But for the moment please remove `token_type_ids` from the inputs of InferenceSession if it doesn't exist in the exported ONNX.

File: src/transformers/models/deberta/configuration_deberta.py
Comment (sam-h-bean) [reply src/transformers/models/deberta/configuration_deberta.py:160]: done

File: src/transformers/models/deberta_v2/configuration_deberta_v2.py
Comment (sam-h-bean) [reply src/transformers/models/deberta_v2/configuration_deberta_v2.py:161]: done

**Comment by sam-h-bean** [comment]
> Looks great, thanks for iterating it!
> 
> It probably makes sense to filter out the invalid inputs and I opened a PR on Optimum. But for the moment please remove `token_type_ids` from the inputs of InferenceSession if it doesn't exist in the exported ONNX.

@JingyaHuang I'm not sure what you mean by this. Do you want something beyond removing the inputs in generate dummy inputs?

**Comment by ChainYo** [comment]
Congratz @sam-h-bean, excellent work. Thanks for adding these configs!!

**Comment by sam-h-bean** [comment]
Can I get a t-shirt 😏 ?

**Pull request merged**
